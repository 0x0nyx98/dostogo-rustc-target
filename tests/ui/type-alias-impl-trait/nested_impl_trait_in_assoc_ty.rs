// edition: 2021

#![feature(impl_trait_in_assoc_type)]

use std::future::Future;

pub struct MemtableLocalStateStore {
    mem_table: MemTable,
}

impl LocalStateStore for MemtableLocalStateStore {
    type IterStream<'a> = impl Sized + 'a where Self: 'a;

    fn iter(&self) -> impl Future<Output = Self::IterStream<'_>> + '_ {
        async move { merge_stream(self.mem_table.iter()) }
    }
}

trait LocalStateStore {
    type IterStream<'a>
    where
        Self: 'a;

    fn iter(&self) -> impl Future<Output = Self::IterStream<'_>> + '_;
}

struct MemTable;

impl MemTable {
    fn iter<'a>(&'a self) -> impl Iterator<Item = &'a ()> {
        std::iter::empty()
    }
}

pub(crate) async fn merge_stream<'a>(mem_table_iter: impl Iterator<Item = &'a ()>) {}

fn main() {}
