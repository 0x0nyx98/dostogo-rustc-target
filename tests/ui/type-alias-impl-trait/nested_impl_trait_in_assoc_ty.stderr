error[E0792]: non-defining opaque type use in defining scope
  --> $DIR/nested_impl_trait_in_assoc_ty.rs:15:9
   |
LL |         async move { merge_stream(self.mem_table.iter()) }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ argument `'_` is not a generic parameter
   |
note: for this opaque type
  --> $DIR/nested_impl_trait_in_assoc_ty.rs:35:1
   |
LL | pub(crate) async fn merge_stream<'a>(mem_table_iter: impl Iterator<Item = &'a ()>) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0792`.
